### ISSUE TITLE
Suggestion: Replace python-ldap with ldap3

### ISSUE THREAD

**Issue by benf710** [django-auth-ldap/django-auth-ldap#325]
# Problem Description

On Windows, it's very cumbersome to install `python-ldap`. Their docs mention to use a pre-built binary but even that can be difficult to find for specific python  versions and  architectures. Pre-built 3rd-party binaries also pose a potential security risk. A better python lib for LDAP has been around for a while, `ldap3`. It's a pure python library and could be a better alternative. I'd like to work on a PR to resolve this issue but before starting I wanted to look for some feedback from the maintainers, if that's even something you would be receptive to and what the best course of action might be.

# Possible Solutions
 - Replacing `python-ldap` with `ldap3` entirely
 - Implementing  a separate backend ex `django_auth_ldap.backend.LDAP3Backend`

Also open to suggestions if you have a better idea for getting around python-ldap pain.

**Comment by francoisfreitag** [comment]
On principle, ldap3 looks pretty attractive. Pure python would be great, both for installation and following the code.
Sadly, it looks like `ldap3` hasn’t been active in since September 2021. It’s still using `travis-ci`, which went read-only (and unusable) since Jan 1st, 2021. I’m unsure whether it is worth spending efforts switching to a provider that looks unmaintained.

For reference, such a proposal has been made in the past with https://github.com/django-auth-ldap/django-auth-ldap/pull/208. It did not proceed further because:
- the change was not tested in production, giving small confidence in the implementation,
- the patch was changing many things at once, making it pretty hard to follow.

**Comment by matmair** [comment]
@francoisfreitag I would be open to help with testing etc. if it helps with the switch. The current requirements outright stop me from using this great library in bigger projects because of the support effort needed for the installation.
Is there still interest in getting #208 or something similar merged or should I just fork #208 and go from there?

**Comment by francoisfreitag** [comment]
Before to spend time on switching, it would be best to see some activity on ldap3. It currently does not look actively maintained.

**Comment by jsoligny** [comment]
There's already an existing fork with LDAP3 : https://github.com/ProbstDJakob/django-auth-ldap. Works fine for me.

**Comment by matmair** [comment]
Thank you for the suggestion @jsoligny. This fork has a bus factor of 0-1; I can not use that in a project a few thousand ppl. use to run their business.

**Comment by johnthagen** [comment]
For others researching this, there is also:

- https://github.com/etianen/django-python3-ldap

that uses `ldap3` under the hood.

**Comment by intgr** [comment]
Not making any promises, but I may be interested in pushing forward efforts to replace `python-ldap` dependency in django-auth-ldap:
1. `python-ldap` is barely and upstream is only interested in supporting Linux.
2. There are no up to date python-ldap binaries on Windows, see https://github.com/python-ldap/python-ldap/issues/511. And even older binaries for Python 3.11 don't seem to work.
3. There are subtle differences in behavior between Linux and macOS (e.g. TLS options) because Apple ships their own fork of LDAP libraries.

My understanding is that `ldap3` is currently the **only** viable alternative. But as pointed out above, `ldap3` itself is also not very actively developed. Most issues go without replies. But at least was at least a typo PR merged 5 days ago! https://github.com/cannatag/ldap3/pull/1033, not sure if this qualifies as "[seeing] some activity on ldap3" 😆

Sounds like we're between a rock and a hard place.

**Comment by intgr** [comment]
Yes, sorry, I meant "barely maintained" :)
